File: iosApp/iosApp/Modules/CurrencySend/CurrencySendInteractor.swift
// Created by web3d4v on 06/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol CurrencySendInteractor: AnyObject {
    var walletAddress: String? { get }
    func defaultCurrency(network: Network) -> Currency
    func balance(currency: Currency, network: Network) -> BigInt
    func networkFees(network: Network) -> [NetworkFee]
}

final class DefaultCurrencySendInteractor {
    private let walletService: WalletService
    private let networksService: NetworksService
    
    init(
        walletService: WalletService,
        networksService: NetworksService
    ) {
        self.walletService = walletService
        self.networksService = networksService
    }
}

extension DefaultCurrencySendInteractor: CurrencySendInteractor {
    
    var walletAddress: String? {
        try? networksService.wallet()?.address().toHexStringAddress().hexString
    }
    
    func defaultCurrency(network: Network) -> Currency {
        walletService.currencies(network: network).first ?? network.nativeCurrency
    }
    
    func balance(currency: Currency, network: Network) -> BigInt {
        walletService.balance(network: network, currency: currency)
    }
    
    func networkFees(network: Network) -> [NetworkFee] {
        networksService.networkFees(network: network)
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencySend/CurrencySendInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.currencySend

import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.legacy.NetworkFee
import com.sonsofcrypto.web3lib.types.toHexStringAddress
import com.sonsofcrypto.web3lib.types.BigInt

interface CurrencySendInteractor {
    val walletAddress: String?
    fun defaultCurrency(network: Network): Currency
    fun balance(currency: Currency, network: Network): BigInt
    fun networkFees(network: Network): List<NetworkFee>
    fun fiatPrice(currency: Currency): Double
}

class DefaultCurrencySendInteractor(
    private val walletService: WalletService,
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
): CurrencySendInteractor {

    override val walletAddress: String? get() =
        networksService.wallet()?.address()?.toHexStringAddress()?.hexString

    override fun defaultCurrency(network: Network): Currency =
        walletService.currencies(network).firstOrNull() ?: network.nativeCurrency

    override fun balance(currency: Currency, network: Network): BigInt =
        walletService.balance(network, currency)

    override fun networkFees(network: Network): List<NetworkFee> =
        networksService.networkFees(network)

    override fun fiatPrice(currency: Currency): Double =
        currencyStoreService.marketData(currency)?.currentPrice ?: 0.toDouble()
}